platform/android/java/lib/src/org/godotengine/godot/input/GodotInputHandler.java

```diff
@@ -368,4 +394,90 @@ private int findJoystickDevice(int device_id) {
 
 		return -1;
 	}
+
+	/**
+	 *
+	 * @param event
+	 * @param isCaptured - must be used and not MOUSE_MODE_CAPTURED because requestPointerCapture is not atomic and can make a race condition on fast mouse movement right after request
+	 */
+	private boolean handleMouseEvent(MotionEvent event, final boolean isCaptured) {
+		if (event.getActionMasked() == MotionEvent.ACTION_HOVER_ENTER ||
+				event.getActionMasked() == MotionEvent.ACTION_HOVER_MOVE ||
+				event.getActionMasked() == MotionEvent.ACTION_HOVER_EXIT) {
+			final int x = Math.round(event.getX());
+			final int y = Math.round(event.getY());
+			final int type = event.getAction();
+			queueEvent(new Runnable() {
+				@Override
+				public void run() {
+					GodotLib.hover(type, x, y);
+				}
+			});
+			return true;
+		} else if (event.getAction() == MotionEvent.ACTION_BUTTON_PRESS || event.getAction() == MotionEvent.ACTION_BUTTON_RELEASE) {
+			final float x = event.getX();
+			final float y = event.getY();
+			final int buttonMask = event.getButtonState();
+			final int action = event.getAction();
+			queueEvent(new Runnable() {
+				@Override
+				public void run() {
+					GodotLib.mouseEvent(action, buttonMask, x, y, isCaptured);
+				}
+			});
+			return true;
+		} else if (event.getAction() == MotionEvent.ACTION_SCROLL) {
+			final float x = event.getX();
+			final float y = event.getY();
+
+			if (event.getAxisValue(MotionEvent.AXIS_VSCROLL) > 0.0f)
+				queueEvent(new Runnable() {
+					@Override
+					public void run() {
+						GodotLib.mouseEvent(MotionEvent.ACTION_BUTTON_PRESS, MOUSE_WHEEL_UP, x, y, isCaptured);
+						GodotLib.mouseEvent(MotionEvent.ACTION_BUTTON_RELEASE, MOUSE_WHEEL_UP, x, y, isCaptured);
+					}
+				});
+			else {
+				queueEvent(new Runnable() {
+					@Override
+					public void run() {
+						GodotLib.mouseEvent(MotionEvent.ACTION_BUTTON_PRESS, MOUSE_WHEEL_DOWN, x, y, isCaptured);
+						GodotLib.mouseEvent(MotionEvent.ACTION_BUTTON_RELEASE, MOUSE_WHEEL_DOWN, x, y, isCaptured);
+					}
+				});
+			}
+			if (event.getAxisValue(MotionEvent.AXIS_HSCROLL) > 0.0f) {
+				queueEvent(new Runnable() {
+					@Override
+					public void run() {
+						GodotLib.mouseEvent(MotionEvent.ACTION_BUTTON_PRESS, MOUSE_WHEEL_RIGHT, x, y, isCaptured);
+						GodotLib.mouseEvent(MotionEvent.ACTION_BUTTON_RELEASE, MOUSE_WHEEL_RIGHT, x, y, isCaptured);
+					}
+				});
+			} else {
+				queueEvent(new Runnable() {
+					@Override
+					public void run() {
+						GodotLib.mouseEvent(MotionEvent.ACTION_BUTTON_PRESS, MOUSE_WHEEL_LEFT, x, y, isCaptured);
+						GodotLib.mouseEvent(MotionEvent.ACTION_BUTTON_RELEASE, MOUSE_WHEEL_LEFT, x, y, isCaptured);
```

Comment:
Use the `GodotLib.scroll` method as defined above and pass the value from `event.getAxisValue(...)` directly instead of interpreting it and passing a constant (the interpretation can be done in the native code).
Since a mouse scroll has the same start and end point, use the same `x` and `y` for the `startX`, `startY` and `endX`, `endY` parameters.

Reply:
I"m not 100% on how to that as mouse scroll up should be added to button state (or should it?) as it works in windows

Reply:
By passing the axis value for the `AXIS_HSCROLL` and `AXIS_VSCROLL` to the native layer, the native layer can decide which is `SCROLL_UP`, `SCROLL_DOWN`, `SCROLL_RIGHT` or `SCROLL_LEFT` using logic similar to what you currently have in the java code.
In addition, you then also getthe value to fill the [InputEventMouseButton#factor](https://docs.godotengine.org/en/3.2/classes/class_inputeventmousebutton.html#class-inputeventmousebutton-property-factor) property.

Reply:
also android onScroll doesn't send anything in `event.getAxisValue(MotionEvent.AXIS_VSCROLL) ` it just doesn't seems right to merge the two

Reply:
I think by button state you mean [InputEventMouseButton#button_index](https://docs.godotengine.org/en/3.2/classes/class_inputeventmousebutton.html#class-inputeventmousebutton-property-button-index). If so that can be done in the native code as I just mentioned.
As a nice side effect, this cleans up the logic in the java code since now only one call to `scroll` is needed to cover all axis.

Reply:
And there are valid values being sent for `event.getAxisValue(MotionEvent.AXIS_VSCROLL)`.
I've tested the logic on a Pixel 2 XL device running Android 10 with a bluetooth mouse, and the value is being sent as expected.

Reply:
by button state I mean the bitmask of all the buttons that are pressed

Reply:
> And there are valid values being sent for event.getAxisValue(MotionEvent.AXIS_VSCROLL).
I've tested the logic on a Pixel 2 XL device running Android 10 with a bluetooth mouse, and the value is being sent as expected.
it just sends 0 in the onScroll

Reply:
> by button state I mean the bitmask of all the buttons that are pressed

These values are irrelevant for scrolling; there are not in the code you have now and there are not in the suggestions I made so I'm not sure why they matter here.

Reply:
``` c++
if (ev->is_pressed()) {
	buttons_state |= button_mask;
} else {
	buttons_state &= ~button_mask;
}```

Reply:
I think you're misunderstanding the feedback that I left for this logic. What I meant is that the entirety of the logic for `ACTION_SCROLL` can be simplified to:
```
if (event.getAction() == MotionEvent.ACTION_SCROLL) {
			final float x = event.getX();
			final float y = event.getY();
                        queueEvent(new Runnable() {
					@Override
					public void run() {
						GodotLib.scroll(event.getSource(), x, y, x, y, event.getAxisValue(MotionEvent.AXIS_HSCROLL), event.getAxisValue(MotionEvent.AXIS_VSCROLL));
					}
				});
}
```

Now in the native code, as you had in the java code there can be logic like
```
int mouse_scroll_button_index =xScroll > 0.0f ? BUTTON_WHEEL_RIGHT : BUTTON_WHEEL...
```
And then use these value for the `button_state`.

Also, I've tested on two devices (both running pure versions of Android 10), and axis values are correctly being sent for `AXIS_VSCROLL` and `AXIS_HSCROLL`.
It could be that your device may be the reason why you're not seeing them, hence why I had mentioned earlier about the need to validate the logic on multiple devices.

Reply:
of course, the problem is gesture on scroll, mouse on scroll works great

Reply:
my current problem is that I needed to call scroll twice one for button down and one for button up, and just calling `input-parse_input_event()` does not seem to work (tried `accumulate_input_event` but it also doesn't work)

Reply:
Doesn't sound correct. Can you show the code where you're calling scroll twice?

Reply:
``` c++
if ((source & SOURCE_MOUSE) == SOURCE_MOUSE) {
	Ref<InputEventMouseButton> ev;
	ev.instance();
	ev->set_position(p_end_pos);;
	ev->set_pressed(true);
	if (h_scroll != 0) {
		Ref<InputEventMouseButton> evd = ev->duplicate();
		evd->set_factor(h_scroll);
		if(h_scroll > 0) {
			evd->set_button_index(BUTTON_WHEEL_RIGHT);
		} else {
			evd->set_button_index(BUTTON_WHEEL_LEFT);
		}
		input->parse_input_event(evd);

		Ref<InputEventMouseButton> evdd = evd->duplicate();
		evd->set_pressed(false);
		input->parse_input_event(evdd);
	}
	if (v_scroll != 0) {
		Ref<InputEventMouseButton> evd = ev->duplicate();
		evd->set_factor(v_scroll);
		if(v_scroll > 0) {
			evd->set_button_index(BUTTON_WHEEL_UP);
		} else {
			evd->set_button_index(BUTTON_WHEEL_DOWN);
		}
		input->parse_input_event(evd);

		Ref<InputEventMouseButton> evdd = evd->duplicate();
		evd->set_pressed(false);
		input->parse_input_event(evdd);
	}
}
```

Reply:
on the old code I used
``` java
if (event.getAxisValue(MotionEvent.AXIS_HSCROLL) > 0.0f) {
				queueEvent(new Runnable() {
					@Override
					public void run() {
						GodotLib.mouseEvent(MotionEvent.ACTION_BUTTON_PRESS, MOUSE_WHEEL_RIGHT, x, y, isCaptured);
						GodotLib.mouseEvent(MotionEvent.ACTION_BUTTON_RELEASE, MOUSE_WHEEL_RIGHT, x, y, isCaptured);
					}
				});
			} else {
			
```

Reply:
The example I gave was not supposed to be used literally. You're supposed to check which between horizontal vs vertical scroll has a larger magnitude and pass that along:
Pseudocode
```
if (abs(v_scroll) >= abs(h_scroll) {
     // create input event for vertical scroll as done in your example
} else {
    // create input event for horizontal scroll as done in your example
}

Reply:
it was not taken literally, as I don't actualy have a way to test hScroll I assume with ball scroller you can get both events at once, that how I did it before (you can check the current code).
but the problem here is not that, look that I call twice for each one of the, one with `set_pressed(true)` and one with `set_pressed(false)` that is where I get the problem

Reply:
OK, I tried doing your changes for about 5 hours or so, it seems to only make more trouble than helping, the only thing that looks better is the onTouchEvent merge on the `ACTION_MOVE`, everything else required an extra hack to work (for example:
``` c++
if(p_button_mask != buttons_state && p_action == ACTION_MOVE) {
	p_action = ((buttons_state & (buttons_state ^ p_button_mask)) == 0) ?
		ACTION_DOWN : ACTION_UP;
}
```
this is needed because `ACTION_UP` happend only on the first mouse button that is pressed)
I think rolling back most of it...

(while currently the only problem is the one mention above)